src/local/room_position.rs

```diff
@@ -19,104 +24,258 @@ use crate::objects::{HasPosition, RoomPosition};
-/// use [`LocalRoomPosition::remote`].
-///
 /// # Serialization
 ///
 /// `LocalRoomPosition` implements both `serde::Serialize` and
 /// `serde::Deserialize`.
 ///
-/// When serializing, it will use the obvious format of `{roomName: String, x:
-/// u32, y: u32}` in "human readable" formats like JSON, and a less obvious
-/// format `{room_x: u32, room_y: u32, x: u32, y: u32}` in "non-human readable"
-/// formats like [`bincode`].
+/// When serializing, it will use the format `{roomName: String, x: u32, y:
+/// u32}` in "human readable" formats like JSON, and will serialize as a single
+/// `i32` in "non-human readable" formats like [`bincode`].
 ///
 /// You can also pass `LocalRoomPosition` into JavaScript using the `js!{}`
 /// macro provided by `stdweb`, or helper methods using the same code like
-/// [`MemoryReference::set`][crate::memory::MemoryReference::set].  It will be
+/// [`MemoryReference::set`][crate::memory::MemoryReference::set]. It will be
 /// serialized the same as in JSON, as an object with `roomName`, `x` and `y`
 /// properties.
 ///
 /// *Note:* serializing using `js!{}` or `MemoryReference::set` will _not_
-/// create a `RoomPosition`, only something with the same properties. Use
-/// `.remote()` if you need a `RoomPosition`.
+/// create a `LocalRoomPosition`, only something with the same properties.
+///
+/// If you need a remote `RoomPosition`, you have two options:
+///
+/// 1.  Use `.remote()`, then send that to JavaScript
+///
+/// 2. Transfer the int from [`LocalRoomPosition::packed_repr`] into JS, and use
+///    the `pos_from_packed` JavaScript function provided by
+///    this library:
+///
+///    ```no_run
+///    # #[macro_use]
+///    # extern crate stdweb;
+///    # use screeps::LocalRoomPosition;
+///    # fn main() {/*
+///    let pos = LocalRoomPosition::new(...)
+///    # */
+///    # let pos: LocalRoomPosition = unimplemented!();
+///    js! {
+///        let pos = pos_from_packed(@{pos.packed_repr()})
+///    };
+///    # }
+///    ```
 ///
 /// [`bincode`]: https://github.com/servo/bincode
-#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+/// [`RoomPosition::local`]: crate::RoomPosition::local
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+#[repr(transparent)]
 pub struct LocalRoomPosition {
-    pub room_name: LocalRoomName,
-    pub x: u32,
-    pub y: u32,
+    /// A bit-packed integer, containing, from highest-order to lowest:
+    ///
+    /// - 1 byte: (room_x) + 128
+    /// - 1 byte: (room_y) + 128
+    /// - 1 byte: x
+    /// - 1 byte: y
+    ///
+    /// For `Wxx` rooms, `room_x = -xx - 1`. For `Exx` rooms, `room_x = xx`.
+    ///
+    /// For `Nyy` rooms, `room_y = -yy - 1`. For `Syy` rooms, `room_y = yy`.
+    ///
+    /// This is the same representation used in the Screeps server, allowing for
+    /// easy translation. Besides the method names and display representation,
+    /// this is the one part of LocalRoomPosition copied directly from the
+    /// engine code.
+    packed: u32,
```

Comment:
I think `i32` would cause a whole other set of conversions when getting coordinates?

The main reason two reasons I chose `u32` were:
-  the bitwise operators act nicely for bitpacking. `i32`'s  `>>` operator will sign-extend, which could be subtly wrong?
- none of the values stored in this bitwise repr are ever negative - even the world position

Somewhat differently, I did u32 over `[u8; 4]` mainly for ease of translating to JS. `packed as i32` should be free, and handle the values correctly.